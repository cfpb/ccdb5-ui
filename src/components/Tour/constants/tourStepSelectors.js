export const DOCUMENT_STEP_SELECTORS = {
  STEP_1: '.content__hero',
  STEP_2: '.m-pagination__btn-next',
  STEP_3: '.saved__search-panel div:nth-child(2)',
  STEP_4: '.back-to-results',
};

export const GEO_STEP_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '.map-shading', label: 'Map Shading' },
  STEP_13: { selector: '#tile-chart-map', label: 'Tile Map' },
  STEP_14: { selector: '.row-chart-section', label: 'Row Charts' },
  STEP_15: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};

export const LIST_COMPLAINTS_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '[data-tour="select-size"]', label: 'Show Number' },
  STEP_13: { selector: '[data-tour="select-sort"]', label: 'Sort Results' },
  STEP_14: {
    selector: '.refine-bar section.m-btn-group',
    label: 'Show Narratives',
  }, // read complaints
  STEP_15: { selector: '.cards-panel', label: 'Results' },
  STEP_16: { selector: '.to-detail', label: 'View Detailed Complaints' }, // complaint id
  STEP_17: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};

export const TRENDS_SELECTORS = {
  STEP_1: { selector: '.content', label: 'Index' },
  STEP_2: {
    selector: 'header.content__hero .m-list--horizontal',
    label: 'Help Links',
  },
  STEP_3: { selector: '.tabbed-navigation', label: 'Views' },
  STEP_4: { selector: '.filter-panel', label: 'Filters' },
  STEP_5: { selector: '.date-filter', label: 'Date Received' },
  STEP_6: {
    selector: '.filter-panel .aggregation.product',
    label: 'Product/Issue',
  },
  STEP_7: {
    selector: '.filter-panel .aggregation button',
    label: 'Show/Hide',
  }, // show/hide filters
  STEP_8: { selector: '.search-bar', label: 'Search Bar' },
  STEP_9: { selector: '#search-summary h2', label: 'Search Summary' },
  STEP_10: {
    selector: '.export-results button.export-btn',
    label: 'Export Data',
  },
  STEP_11: {
    selector: '.export-results button.print-preview',
    label: 'Print Page',
  }, // print button
  STEP_12: { selector: '[data-tour="select-lens"]', label: 'Aggregate By' },
  STEP_13: {
    selector: '[data-tour="select-interval"]',
    label: 'Date Interval',
  },
  STEP_14: { selector: '.row-chart-section', label: 'Row Charts' },
  STEP_15: { selector: '.tabbed-navigation', label: 'Tabbed Navigation' },
};
