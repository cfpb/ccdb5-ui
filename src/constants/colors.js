module.exports = Object.freeze({
  BriteCharts: {
    dark: '#118935',
    regular: '#20aa3f',
    medium: '#ADDC91',
    light: '#C7E5B3',
  },
  // if you modify anything here,
  // make sure that the following colors are in sync
  // Line Chart
  // Area Chart
  // Stacked Area Chart
  // External Tooltip/Legend
  DataLens: {
    0: '#2cb34a', // 0
    1: '#addc91', // 1
    2: '#0072ce', // 4
    3: '#96c4ed', // 5
    4: '#257675', // 2
    5: '#9ec4c3', // 3
    6: '#254b87',
    7: '#9daecc',
    8: '#b4267a',
    9: '#dc9cbf',
    10: '#a2a3a4',
  },
});
