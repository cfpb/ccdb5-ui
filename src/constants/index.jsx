// Internal triggers
export const API_CALLED = 'API_CALLED'
export const MODAL_HID = 'MODAL_HID'
export const MODAL_SHOWN = 'MODAL_SHOWN'

// Modal Dialogs
export const MODAL_TYPE_DATA_EXPORT = 'MODAL_TYPE_DATA_EXPORT'
export const MODAL_TYPE_MORE_ABOUT = 'MODAL_TYPE_MORE_ABOUT'

// Useful constants
export const NARRATIVE_SEARCH_FIELD = 'complaint_what_happened'

// query manager flags
// These constants control how the query manager works
export const REQUERY_ALWAYS = 'REQUERY_ALWAYS'
export const REQUERY_HITS_ONLY = 'REQUERY_HITS_ONLY'
// default if not specified
export const REQUERY_NEVER = 'REQUERY_NEVER'

export const knownFilters = [
  'company',
  'company_public_response',
  'company_response',
  'consumer_consent_provided',
  'consumer_disputed',
  'issue',
  'product',
  'state',
  'submitted_via',
  'tag',
  'timely',
  'zip_code'
]

export const dateFilters = [
  'company_received_max',
  'company_received_min',
  'date_received_max',
  'date_received_min'
]

export const flagFilters = [
  'has_narrative'
]

export const SLUG_SEPARATOR = 'â€¢'

// Keyboard
export const VK_DOWN = 40
export const VK_ENTER = 14
export const VK_ESCAPE = 27
export const VK_RETURN = 13
export const VK_TAB = 9
export const VK_UP = 38

/// ////////////////////////////////////////////////////////////////////////////
// JSON Objects

export const THESE_UNITED_STATES = {
  AA: 'Armed Forces Americas',
  AE: 'Armed Forces Europe',
  AL: 'Alabama',
  AK: 'Alaska',
  AS: 'American Samoa',
  AP: 'Armed Forces Pacific',
  AZ: 'Arizona',
  AR: 'Arkansas',
  CA: 'California',
  CO: 'Colorado',
  CT: 'Connecticut',
  DE: 'Delaware',
  DC: 'District Of Columbia',
  FM: 'Federated States Of Micronesia',
  FL: 'Florida',
  GA: 'Georgia',
  GU: 'Guam',
  HI: 'Hawaii',
  ID: 'Idaho',
  IL: 'Illinois',
  IN: 'Indiana',
  IA: 'Iowa',
  KS: 'Kansas',
  KY: 'Kentucky',
  LA: 'Louisiana',
  ME: 'Maine',
  MH: 'Marshall Islands',
  MD: 'Maryland',
  MA: 'Massachusetts',
  MI: 'Michigan',
  MN: 'Minnesota',
  MS: 'Mississippi',
  MO: 'Missouri',
  MT: 'Montana',
  NE: 'Nebraska',
  NV: 'Nevada',
  NH: 'New Hampshire',
  NJ: 'New Jersey',
  NM: 'New Mexico',
  NY: 'New York',
  NC: 'North Carolina',
  ND: 'North Dakota',
  MP: 'Northern Mariana Islands',
  OH: 'Ohio',
  OK: 'Oklahoma',
  OR: 'Oregon',
  PW: 'Palau',
  PA: 'Pennsylvania',
  PR: 'Puerto Rico',
  RI: 'Rhode Island',
  SC: 'South Carolina',
  SD: 'South Dakota',
  TN: 'Tennessee',
  TX: 'Texas',
  UT: 'Utah',
  VT: 'Vermont',
  VI: 'Virgin Islands',
  VA: 'Virginia',
  WA: 'Washington',
  WV: 'West Virginia',
  WI: 'Wisconsin',
  WY: 'Wyoming'
}

export const TILE_MAP_STATES = [
  'AK',
  'AL',
  'AR',
  'AZ',
  'CA',
  'CO',
  'CT',
  'DC',
  'DE',
  'FL',
  'GA',
  'HI',
  'IA',
  'ID',
  'IL',
  'IN',
  'KS',
  'KY',
  'LA',
  'MA',
  'MD',
  'ME',
  'MI',
  'MN',
  'MO',
  'MS',
  'MT',
  'NC',
  'ND',
  'NE',
  'NH',
  'NJ',
  'NM',
  'NV',
  'NY',
  'OH',
  'OK',
  'OR',
  'PA',
  'RI',
  'SC',
  'SD',
  'TN',
  'TX',
  'UT',
  'VA',
  'VT',
  'WA',
  'WI',
  'WV',
  'WY'
]
