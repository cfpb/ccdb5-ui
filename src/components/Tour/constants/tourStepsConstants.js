import { MODE_LIST, MODE_MAP, MODE_TRENDS } from '../../../constants';

import {
  GEO_STEP_SELECTORS as GEO,
  LIST_COMPLAINTS_SELECTORS as LIST,
  TRENDS_SELECTORS as TRENDS,
} from './tourStepSelectors';

const generateLinkText = (index, linkObject) =>
  `<li><button class="a-btn a-btn--link" onclick="document.querySelectorAll('.introjs-bullets li a')[${
    index + 1
  }].click()">` +
  linkObject[1].label +
  '</button></li>';

let geoIndex = '';

// we exclude the first entry so we can have Index appear as the title
Object.entries(GEO)
  .slice(1)
  .forEach((value, key) => {
    geoIndex += generateLinkText(key, value);
  });

let listIndex = '';
Object.entries(LIST)
  .slice(1)
  .forEach((value, key) => {
    listIndex += generateLinkText(key, value);
  });

let trendsIndex = '';
Object.entries(TRENDS)
  .slice(1)
  .forEach((value, key) => {
    trendsIndex += generateLinkText(key, value);
  });

export const TOUR_STEPS = {
  [MODE_MAP]: [
    {
      element: GEO.STEP_1.selector,
      intro:
        '<div>' +
        '<div class="left">' +
        'Welcome to the Consumer Complaint Database!' +
        '<br /><br />' +
        'Complaints the CFPB sends to companies for response are published in the Consumer Complaint Database after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.' +
        '<br /><br />' +
        'Complaints the CFPB refers to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the database. This database is not a statistical sample of consumers’ experiences in the marketplace.' +
        '</div>' +
        '<div class="right">' +
        '<h4>Index</h4>' +
        '<ul>' +
        geoIndex +
        '</ul></div>' +
        '</div>',
      tooltipClass: 'wide',
    },
    {
      element: GEO.STEP_2.selector,
      intro: 'These links provide more information about this database.',
    },
    {
      element: GEO.STEP_3.selector,
      intro:
        'You can change how to view complaint data by selecting the three views we have available — Trends, List, and Map.' +
        '<br /><br />' +
        'You are currently on the Map view, which displays complaints by state.' +
        '<br /><br />' +
        'To read individual complaints, select List view' +
        '<br /><br />' +
        'To visualize complaint data with product and issue breakdowns, select Trends view',
      position: 'top',
    },
    {
      element: GEO.STEP_4.selector,
      intro:
        'The “Filter results by” panel lets you filter complaint data. To clear filters, uncheck individual selections in the filter panel or to select “Clear all filters” under the Search bar.',
    },
    {
      element: GEO.STEP_5.selector,
      intro:
        'You can adjust the date range of the complaint data showed using the “Date CFPB received the complaint” filter. <br /> The default date range is three years from today’s date.',
    },
    {
      element: GEO.STEP_6.selector,
      intro:
        'You can filter complaints by the product and issue the consumer selected when they submitted their complaint.' +
        '<br /><br />' +
        'Expand a product or issue to see its sub-products or sub-issues, where available. ',
    },
    {
      element: GEO.STEP_7.selector,
      intro:
        'The blue show/hide icon allows you to expand or collapse specific filters.',
    },
    {
      element: GEO.STEP_8.selector,
      intro:
        'Use the “Search within” bar to find specific words or phrases in complaints. By default, this will search “All data”, but you can change this using the gray dropdown menu.' +
        '<br /> <br />' +
        'Click “Show advanced search tips” to see some advanced ways to refine your word search.',
    },
    {
      element: GEO.STEP_9.selector,
      intro:
        'As you apply filters and search terms, this will display how many complaints are included out of the total number of complaints published in the database.',
    },
    {
      element: GEO.STEP_10.selector,
      intro:
        'You can export full or filtered results to CSV or JSON by selecting “Export data”.',
    },
    {
      element: GEO.STEP_11.selector,
      intro: 'Select “Print” to generate a pdf of the current page.',
    },
    {
      element: GEO.STEP_12.selector,
      intro:
        'Map shading defaults to total complaints, but you can change this to show complaints per 1,000 population in that state.',
    },
    {
      element: GEO.STEP_13.selector,
      intro:
        'Hover over a tile on the map to view complaint information specific to that state. You can filter to see data specific to a state by selecting the state or using the filter panel.',
    },
    {
      element: GEO.STEP_14.selector,
      intro:
        'View complaint volume by product in the bar chart. Use the blue arrow by each product to view breakdowns by sub-products.' +
        '<br /><br />Once the product is expanded, select the “Visualize” link as shown below, to view trends on sub-products and issues:',
    },
    {
      element: GEO.STEP_15.selector,
      intro:
        'To learn how to use Trends and List views, select the view and then start the tour.',
    },
  ],
  [MODE_LIST]: [
    {
      element: LIST.STEP_1.selector,
      intro:
        '<div>' +
        '<div class="left">' +
        'Welcome to the Consumer Complaint Database!' +
        '<br /><br />' +
        'Complaints the CFPB sends to companies for response are published in the Consumer Complaint Database after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.' +
        '<br /><br />' +
        'Complaints the CFPB refers to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the database. This database is not a statistical sample of consumers’ experiences in the marketplace.' +
        '</div>' +
        '<div class="right">' +
        '<h4>Index</h4>' +
        '<ul>' +
        listIndex +
        '</ul></div>' +
        '</div>',
      tooltipClass: 'wide',
    },
    {
      element: LIST.STEP_2.selector,
      intro: 'These links provide more information about this database.',
    },
    {
      element: LIST.STEP_3.selector,
      intro:
        'You can change how to view complaint data by selecting the three views we have available — Trends, List, and Map.' +
        '<br /><br />' +
        'You are currently on the List view, which shows individual complaints.' +
        '<br /><br />' +
        'To visualize complaint data with product and issue breakdowns, select Trends view.' +
        '<br /><br />' +
        'To view complaints by state, select Map view.',
    },
    {
      element: LIST.STEP_4.selector,
      intro:
        'The “Filter results by” panel lets you filter complaint data. To clear filters, uncheck individual selections in the filter panel or to select “Clear all filters” under the Search bar.',
    },
    {
      element: LIST.STEP_5.selector,
      intro:
        'You can adjust the date range of the complaint data showed using the “Date CFPB received the complaint” filter. <br /> The default date range is three years from today’s date.',
    },
    {
      element: LIST.STEP_6.selector,
      intro:
        'You can filter complaints by the product and issue the consumer selected when they submitted their complaint.' +
        '<br /><br />' +
        'Expand a product or issue to see its sub-products or sub-issues, where available. ',
    },
    {
      element: LIST.STEP_7.selector,
      intro:
        'The blue show/hide icon allows you to expand or collapse specific filters.',
    },
    {
      element: LIST.STEP_8.selector,
      intro:
        'Use the “Search within” bar to find specific words or phrases in complaints. By default, this will search “All data”, but you can change this using the gray dropdown menu.' +
        '<br /> <br />' +
        'Click “Show advanced search tips” to see some advanced ways to refine your word search.',
    },
    {
      element: LIST.STEP_9.selector,
      intro:
        'As you apply filters and search terms, this will display how many complaints are included out of the total number of complaints published in the database.',
    },
    {
      element: LIST.STEP_10.selector,
      intro:
        'You can export full or filtered results to CSV or JSON by selecting “Export data”.',
    },
    {
      element: LIST.STEP_11.selector,
      intro: 'Select “Print” to generate a pdf of the current page.',
    },
    {
      element: LIST.STEP_12.selector,
      intro: 'Select the number of complaints to view per page here.',
    },
    {
      element: LIST.STEP_13.selector,
      intro:
        'Sort complaints in the list view by “Newest to oldest,” “Oldest to newest,” “Relevance,” and “Relevance (asc).”' +
        '<br /><br />' +
        'Relevancy are based on any applied search terms.',
    },
    {
      element: LIST.STEP_14.selector,
      intro:
        'You can see all complaints or focus on those with consumer complaint narratives.' +
        '<br /><br />' +
        'We publish the consumer’s narrative description of what happened from their complaint if the consumer opts to share it publicly and after taking steps to remove personal information.',
    },
    {
      element: LIST.STEP_15.selector,
      intro:
        'These are your search results. As you apply filters and search terms, your results will display here. This view shows a preview of the complaint.',
    },
    {
      element: LIST.STEP_16.selector,
      intro:
        'Click the blue ID number to see more details about a particular complaint.',
    },
    {
      element: LIST.STEP_17.selector,
      intro:
        'To learn how to use Map and Trend views, select the view and then start the tour.',
    },
  ],
  [MODE_TRENDS]: [
    {
      element: TRENDS.STEP_1.selector,
      intro:
        '<div>' +
        '<div class="left">' +
        'Welcome to the Consumer Complaint Database!' +
        '<br /><br />' +
        'Complaints the CFPB sends to companies for response are published in the Consumer Complaint Database after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first.' +
        '<br /><br />' +
        'Complaints the CFPB refers to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the database. This database is not a statistical sample of consumers’ experiences in the marketplace.' +
        '</div>' +
        '<div class="right">' +
        '<h4>Index</h4>' +
        '<ul>' +
        trendsIndex +
        '</ul></div>' +
        '</div>',
      tooltipClass: 'wide first-step',
    },
    {
      element: TRENDS.STEP_2.selector,
      intro: 'These links provide more information about this database.',
    },
    {
      element: TRENDS.STEP_3.selector,
      intro:
        'You can change how to view complaint data by selecting the three views we have available—Trends, List, and Map.' +
        '<br /><br />' +
        'You are currently on the Trends view, which visualizes complaint data using a trend line with product and issue breakdowns.' +
        '<br /><br />' +
        'To read individual complaints, select List view.' +
        '<br /><br />' +
        'To view complaints by state, select Map view.',
      position: 'top',
    },
    {
      element: TRENDS.STEP_4.selector,
      intro:
        'The “Filter results by” panel lets you filter complaint data. To clear filters, uncheck individual selections in the filter panel or to select “Clear all filters” under the Search bar.',
    },
    {
      element: TRENDS.STEP_5.selector,
      intro:
        'You can adjust the date range of the complaint data showed using the “Date CFPB received the complaint” filter. <br /> The default date range is three years from today’s date.',
    },
    {
      element: TRENDS.STEP_6.selector,
      intro:
        'You can filter complaints by the product and issue the consumer selected when they submitted their complaint.' +
        '<br /><br />' +
        'Expand a product or issue to see its sub-products or sub-issues, where available. ',
    },
    {
      element: TRENDS.STEP_7.selector,
      intro:
        'The blue show/hide icon allows you to expand or collapse specific filters.',
    },
    {
      element: TRENDS.STEP_8.selector,
      intro:
        'Use the “Search within” bar to find specific words or phrases in complaints. By default, this will search “All data”, but you can change this using the gray dropdown menu.' +
        '<br /> <br />' +
        'Click “Show advanced search tips” to see some advanced ways to refine your word search.',
    },
    {
      element: TRENDS.STEP_9.selector,
      intro:
        'As you apply filters and search terms, this will display how many complaints are included out of the total number of complaints published in the database.',
    },
    {
      element: TRENDS.STEP_10.selector,
      intro:
        'You can export full or filtered results to CSV or JSON by selecting “Export data”.',
    },
    {
      element: TRENDS.STEP_11.selector,
      intro: 'Select “Print” to generate a pdf of the current page.',
    },
    {
      element: TRENDS.STEP_12.selector,
      intro:
        'Select the dropdown to aggregate the data by product or by the company to which the CFPB sent the complaint for response.',
    },
    {
      element: TRENDS.STEP_13.selector,
      intro:
        'Select this dropdown to change the time interval displayed on the trend line.',
    },
    {
      element: TRENDS.STEP_14.selector,
      intro:
        'View complaint volume by product in the bar chart. Use the blue arrow by each product to view breakdowns by sub-products.' +
        '<br /><br />Once the product is expanded, select the “Visualize” link as shown below, to view trends on sub-products and issues:',
    },
    {
      element: TRENDS.STEP_15.selector,
      intro:
        'To learn how to use List and Map views, select the view and then start the tour.',
      position: 'top',
    },
  ],
};
