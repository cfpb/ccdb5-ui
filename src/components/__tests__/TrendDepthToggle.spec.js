import configureMockStore from 'redux-mock-store';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import ReduxTrendDepthToggle, {
  TrendDepthToggle,
  mapDispatchToProps,
  mapStateToProps,
} from '../Trends/TrendDepthToggle';
import React from 'react';
import renderer from 'react-test-renderer';
import { REQUERY_ALWAYS } from '../../constants';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';

/**
 *
 * @param {object} root0 - Root state
 * @param {number} root0.cbIncrease - Amount of depth increase
 * @param {boolean} root0.cbReset - Rest the depth?
 * @param {number} root0.diff - Difference
 * @param {number} root0.queryCount - Query count
 * @param {number} root0.resultCount - Query count
 * @returns {React.Component} - Trends Depth Toggle component
 */
function setupEnzyme({ cbIncrease, cbReset, diff, queryCount, resultCount }) {
  return shallow(
    <TrendDepthToggle
      diff={diff}
      increaseDepth={cbIncrease}
      lens="Product"
      depthReset={cbReset}
      queryCount={queryCount}
      resultCount={resultCount}
      hasToggle={true}
    />,
  );
}

/**
 *
 * @param {object} root0 - Root state
 * @param {string} root0.focus - The focus
 * @param {string} root0.lens - The lens
 * @param {object} root0.productAggs - Product aggs
 * @param {object} root0.productResults - Product results
 * @returns {Function} - Trends Depth Toggle component renderer
 */
function setupSnapshot({ focus, lens, productAggs, productResults }) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    aggs: {
      product: productAggs,
    },
    query: {
      focus,
      lens,
    },
    trends: {
      results: {
        product: productResults,
      },
    },
  });

  return renderer.create(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ReduxTrendDepthToggle />
      </IntlProvider>
    </Provider>,
  );
}

describe('component:TrendDepthToggle', () => {
  let params;

  beforeEach(() => {
    params = {
      focus: '',
      lens: '',
      productAggs: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
      productResults: [
        { name: 'a', visible: true },
        { name: 'b', visible: true },
        { name: 'c', visible: true },
        { name: 'd', visible: true },
        { name: 'e', visible: true },
        { name: 'f', visible: true },
        { name: 'g', visible: true },
        { name: 'h', visible: true },
      ],
    };
  });

  it('does not render when Focus', () => {
    params.focus = 'A focus item';
    const target = setupSnapshot(params);
    const tree = target.toJSON();
    expect(tree).toBeNull();
  });

  it('does not render lens is not Product', () => {
    params.lens = 'Cannot See';
    const target = setupSnapshot(params);
    const tree = target.toJSON();
    expect(tree).toBeNull();
  });

  it('renders Product view more without crashing', () => {
    params.lens = 'Product';
    const target = setupSnapshot(params);
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Product view less without crashing', () => {
    params.lens = 'Product';
    params.productAggs = [1, 2, 3, 4, 5];
    params.productResults = [
      { name: 'a', visible: true, isParent: true },
      { name: 'b', visible: true, isParent: true },
      { name: 'c', visible: true, isParent: true },
      { name: 'd', visible: true, isParent: true },
      { name: 'e', visible: true, isParent: true },
      { name: 'f', visible: true, isParent: true },
      { name: 'g', visible: true, isParent: true },
      { name: 'h', visible: true, isParent: true },
    ];
    const target = setupSnapshot(params);
    const tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('buttons', () => {
    let cbIncrease = null;
    let cbReset = null;
    let target;

    beforeEach(() => {
      cbIncrease = jest.fn();
      cbReset = jest.fn();
    });

    it('increaseDepth is called when the increase button is clicked', () => {
      target = setupEnzyme({
        cbIncrease,
        cbReset,
        diff: 1000,
        resultCount: 5,
      });
      const prev = target.find('#trend-depth-button');
      prev.simulate('click');
      expect(cbIncrease).toHaveBeenCalledWith(1000);
    });

    it('reset depth is called when the reset button is clicked', () => {
      target = setupEnzyme({
        cbIncrease,
        cbReset,
        diff: 0,
        queryCount: 10,
        resultCount: 10,
      });
      const prev = target.find('#trend-depth-button');
      prev.simulate('click');
      expect(cbReset).toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    it('hooks into changeDepth', () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).increaseDepth(13);
      expect(dispatch.mock.calls).toEqual([
        [
          {
            meta: {
              requery: REQUERY_ALWAYS,
            },
            type: 'query/changeDepth',
            payload: { depth: 18 },
          },
        ],
      ]);
    });

    it('hooks into resetDepth', () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).depthReset();
      expect(dispatch.mock.calls).toEqual([
        [
          {
            meta: {
              requery: REQUERY_ALWAYS,
            },
            type: 'query/resetDepth',
          },
        ],
      ]);
    });
  });

  describe('mapStateToProps', () => {
    it('maps state and props', () => {
      const state = {
        aggs: {
          product: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
        },
        query: {
          focus: '',
          lens: 'Product',
        },
        trends: {
          results: {
            product: [
              { name: 'a', visible: true },
              {
                name: 'b',
                visible: true,
              },
              { name: 'c', visible: true },
              { name: 'd', visible: true },
              { name: 'e', visible: true },
              { name: 'f', visible: true },
              { name: 'g', visible: true },
              { name: 'h', visible: true },
            ],
          },
        },
      };
      const actual = mapStateToProps(state);
      expect(actual).toEqual({
        diff: 11,
        queryCount: 11,
        resultCount: 0,
        hasToggle: true,
      });
    });

    describe('when lens = Company', () => {
      let state;
      beforeEach(() => {
        state = {
          aggs: {},
          query: {
            focus: '',
            lens: 'Company',
            company: [
              'I',
              'I',
              'III',
              'IV',
              'V',
              'VI',
              'VII',
              'VIII',
              'IX',
              'X',
              'XI',
            ],
          },
          trends: {
            results: {
              company: [
                { name: 'a', visible: true },
                { name: 'b', visible: true },
                { name: 'c', visible: true },
                { name: 'd', visible: true },
                { name: 'e', visible: true },
                { name: 'f', visible: true },
                { name: 'g', visible: true },
                { name: 'h', visible: true },
                { name: 'i', visible: true },
                { name: 'j', visible: true },
              ],
            },
          },
        };
      });

      it('caps the maximum number of companies at 10', () => {
        const actual = mapStateToProps(state);
        expect(actual).toEqual({
          diff: 10,
          queryCount: 11,
          resultCount: 0,
          hasToggle: true,
        });
      });

      it('shows the toggle when results < 10', () => {
        state.trends.results.company.splice(4, 5);

        const actual = mapStateToProps(state);
        expect(actual).toEqual({
          diff: 10,
          queryCount: 11,
          resultCount: 0,
          hasToggle: true,
        });
      });

      it('hides the toggle when only parent count < 5 item', () => {
        state.query.company = ['a'];
        state.trends.results.company = [
          { name: 'a', visible: true, isParent: true },
          { name: 'b', visible: true, isParent: false },
          { name: 'c', visible: true, isParent: false },
          { name: 'd', visible: true, isParent: false },
          { name: 'e', visible: true, isParent: false },
          { name: 'f', visible: true, isParent: false },
          { name: 'g', visible: true, isParent: false },
          { name: 'h', visible: true, isParent: false },
          { name: 'i', visible: true, isParent: false },
          { name: 'j', visible: true, isParent: false },
        ];

        const actual = mapStateToProps(state);
        expect(actual).toEqual({
          diff: 0,
          queryCount: 1,
          resultCount: 1,
          hasToggle: false,
        });
      });
    });
  });
});
