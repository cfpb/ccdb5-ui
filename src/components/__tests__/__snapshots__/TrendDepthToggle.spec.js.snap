// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:TrendDepthToggle renders Product view less without crashing 1`] = `
<div
  className="trend-depth-toggle"
>
  <button
    className="a-btn a-btn__link"
    id="trend-depth-button"
    onClick={[Function]}
  >
    <span
      className="minus"
    />
    Show less
  </button>
</div>
`;

exports[`component:TrendDepthToggle renders Product view more without crashing 1`] = `
<div
  className="trend-depth-toggle"
>
  <button
    className="a-btn a-btn__link"
    id="trend-depth-button"
    onClick={[Function]}
  >
    <span
      className="plus"
    />
    Show more
  </button>
</div>
`;
